### PR TITLE
Add initialize with partial Table or Column Schema

### LINKED ISSUE #965: Add a way of initializing Woodwork with partial Table or Column Schemas
As a user, I wish I could initialize Woodwork on a DataFrame with a partial schema in a way that would perform inference on the remaining columns and be able to update the dtype of the columns that did have typing information defined.

Currently, there are two ways of initializing Woodwork on a DataFrame:
1. With all the woodwork parameters passed in separately - this allows for specifying partial typing information and inferring on the columns without typing information
2. With a `TableSchema` - requires all typing information be present and correct

Recently, there've been a few places where we needed to use the first initialization option when a significant portion of a DataFrame's typing information is known, but either the dtypes of the DataFrame may have changed or we don't have all of the typing information for all of the columns in the DataFrame so the second approach wasn't an option (woodwork's concat function, and featuretools' `EntitySet.add_last_time_indexes` are two that come to mind). 

The first initialization option makes this kind of initialization from table to table with partial typing information possible, but it is error-prone if a parameter to Woodwork initialization is forgotten or if a parameter should be deepcopied. It also adds more places that we need to remember to update any time Woodwork initialization changes.

Two options for this new method of initialization are as follows:
- Init with a dictionary mapping `col_name -> ColumnSchema` 
  - the main problem with this approach is that table-wide parameters like index and table metadata would still need to get passed in separately.
- Init with a subset `TableSchema`

Both of these would confirm that every column was actually present 

Once we're adding a third way of initializing Woodwork, it's probably too much to have all the logic handled within `df.ww.init`. It might be beneficial to at least have this in a separate method, but we can also consider having three methods:
- `df.ww.init` - pass individual parameters; anything not specified gets inferred.
- `df.ww.init_with_full_schema` - pass full table schema
- `df.ww.init_with_partial_schema` - pass partial schema; anything not specified gets inferred.

### PR DESCRIPTION
**Pull request by tuethan1999**
- creation of init_with_partial_schema, init_with_full_schema, and init
- Closes #965

### CODE REVIEW

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:51]: Didn't really think about this till now, but I wonder we might need to keep `init(schema=schema)` to avoid adding a breaking change. We could call `self.init_with_full_schema(schema=None)` so that it matches the current behavior by just using the schema, assuming it's relevant for the whole table, and ignoring any other parameters.

We'd probably want to add a warning that tells users that they should go through the full or partial schema methods, but I'm thinking that we should avoid the breaking change here if possible.

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py]: Might be misreading/misunderstanding how the types work, but do we want the types for logical_types to be `Optional[dict[str, Union[str, LogicalType]]]`?

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py]: These params that can take in a dictionary where they keys are column names should allow the column names to be anything that can be a pandas column name (so it allows anything hashable like numbers). Something I think could be incredibly useful for us would be to create a `column_name` type alias for these params. So we could say `Optional[dict[column_name, dict]]` and `column_name` would be something like `Union[str, number, tuple]`

@davesque is there a way with python's types to make some sort of type alias? (or maybe the question should be: Is that an advisable thing to do even if it's possible)

File: woodwork/table_accessor.py
Comment (tamargrey) [reply woodwork/table_accessor.py]: side note: So awesome to see the types here!

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:127]: if I'm understanding the goal for this function correctly, you should be able to do `{**right, **left}`, and that will create a new dictionary object that will contain the `right` and `left` items, with any duplicate keys giving preference to `left`

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:662]: I believe `new_schema` will be the full schema for `new_df` since the subset used in line 656 is the same as what was used to create `new_df`, so `init_with_full_schema` may be more clear here

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:150]: Should these objects that may be coming from another DataFrame's schema be deepcopied before being passed into a new DataFrame at ww init?

Basically, my fear is that using the same `table_metadata` object that's coming from a `schema` means that any changes to the resulting DataFrame's woodwork metadata would get propogated back to the DataFrame where this metadata originally came from

**Review by tamargrey — COMMENTED** [review]
Very excited about getting this functionality in Woodwork! I think it'd be good to spend a little time making sure this api can be used for the various use cases that have come up (should all be referenced by the issue) in case we need to create a follow-up issues that gets us these use cases:
- woodwork concat_columns
- featuretools' add_last_time_indexes
- woodwork's `to_frame`
- evalML's use case for retaining as much typing info as possible

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py:145]: I added a `test_init_with_partial_schema_metadata_deep_copy` to show that changes don't propagate

File: woodwork/table_accessor.py
Comment (rwedge) [inline woodwork/table_accessor.py:54]: How difficult would it be to determine if a passed in schema was partial or not, and call the appropriate init method accordingly?

Comment (tuethan1999) [reply woodwork/table_accessor.py]: not very, just threw in a line that checks if all the dataframe columns are in the schema

File: woodwork/tests/accessor/test_table_accessor.py
Comment (tamargrey) [inline woodwork/tests/accessor/test_table_accessor.py:2570]: I think this test wouldn't catch the deepcopy behavior that we need to avoid.

Setting `test_df.ww._schema.metadata` with a completely new dictionary object (line 2569)  will only also change `schema`'s metadata if `schema` and `test_df.ww._schema` are the same `TableSchema` object.

This test needs to update the `test_df.ww._schema.metadata` dictionary that's already there. Adding a new entry will, I believe, test if a shallow copy _of the dictionary_ was performed (if both metadata dicts are the same object). But to confirm that a deep copy has occurred, we'll need to update a mutable value in the metadata dictionary (both lists, for example, would update if the lists are the same object). 

And example of this kind of test is `test_dataframe_methods_on_accessor_new_schema_object` or `test_iloc_table_does_not_propagate_changes_to_data`

**Review by freddyaboulton — COMMENTED** [review]
Is there a shorthand for "Init with my old schema but redo inference on the logical types for some columns?"

In EvalML, we want to eventually get to a point where our components preserve the ww information as much as possible throughout the pipeline. One of the hurdles we face is that some of our components, e.g. Imputers, StandardScaler, modify the data in existing columns as opposed to just creating new features from old columns. The transformations done by these components usually "break" the typing info, e.g. an int gets converted to  a double.

EvalML currently has a janky util function for at least preserving the logical types in the columns that did not get transformed but we'd like to preserve more info (like semantic tags and indices) using `init_with_partial_schema`. 

I'm trying to see how that would look and I think the current implementation does not work 100% for our use case. I'm not requesting changes or anything on this PR, but I'd like your guys' thoughts on how we can achieve what we're after. Maybe there's already an easier way in ww to do it.

The use case is a component to scale numeric columns. This is a minimal repro:

```python
import woodwork as ww
from sklearn.preprocessing import StandardScaler
import pandas as pd

df = pd.DataFrame({"a": [1, 2, 3],
                   "b": [4, 5, 6],
                   "c": ["a", "b", "c"]})
df.ww.init(logical_types={"a": "Integer", "b": "Integer", "c": "Categorical"},
           semantic_tags={"a": ["column a"], "b": ["column b"]})

def scale_numerics(df):
    numeric = df.ww.select(["numeric"])
    scaled = pd.DataFrame(StandardScaler().fit_transform(numeric), columns=numeric.columns)
    return ww.concat_columns([df.ww.drop(list(numeric.columns)), scaled])
```

I can't pass the old schema to `init_with_partial_schema` because then the scaled columns get converted to ints and that's not what we want:

image

I can't "subset" the schema for the non-numeric columns because then some ww info would be lost, in this case the semantic tags for the numeric columns:

image


What I can do is specify the logical types for the non-numeric columns and manually pass in as much as the schema information as possible:

image

I'm wondering if there's a short-hand for the third "solution". Having us do it manually in all our components could lead to bugs and it's possible my "solution" does not carry over all of the relevant schema information, (like the origin field?)

**Comment by tuethan1999** [comment]
> I'm wondering if there's a short-hand for the third "solution". Having us do it manually in all our components could lead to bugs and it's possible my "solution" does not carry over all of the relevant schema information, (like the origin field?)
@freddyaboulton we could add a `force_infer_logical_types` argument

File: woodwork/tests/accessor/test_table_accessor.py
Comment (tuethan1999) [reply woodwork/tests/accessor/test_table_accessor.py:2570]: you're right!  I've changed it to copy the same things that `_get_subset_schema` copies

**Comment by freddyaboulton** [comment]
@tuethan1999 I think that would work! Would the api be `force_infer_logical_types=["Integer", "Boolean"]` or would it infer all logical types again

**Comment by tuethan1999** [comment]
@freddyaboulton I think the api would be more along the lines `force_infer_logical_types=['b', 'c']` where the iterables arguments are column names

**Comment by freddyaboulton** [comment]
@tuethan1999 Sounds good!

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:169]: Not totally sure I'm following what's happening here... But if I understand what the behavior here is:

`use_standard_tags` will be `True` by default, so if that's the case, we assume that that's just the default (and not that the user wanted all columns to use `True`), so we overwrite those values with whatever was in the Schema. This will contradict what the docstring says about the `schema` arg and the `use_standard_tags` arg if the Schema being passed in has some columns as `False`.

What if someone does want to set all of them to True? Not sure that it'd be expected to suddenly have the schema overwrite. I think we might need the default to be None here. That way a user can specify `True` and have it mean `True`.

Can we add some tests that cover this behavior?

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:735]: I know we weren't doing this before, but this seems like a good candidate for `validate=False`

File: woodwork/table_accessor.py
Comment (tamargrey) [reply woodwork/table_accessor.py:49]: In general, we're adding a lot of new functionality with woodwork init here, and I think it could really benefit from an in-depth explanation somewhere. Maybe it makes sense to create an issue to make a guide on the different ways to initialize woodwork.

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py]: I think it's better practice to have it separated out. I separated out the function from the init method and added a docstring to make it clear what it's purpose was

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py]: I think I meant to set the default value for unspecified use_standard_tags columns. I removed this.

File: docs/source/release_notes.rst
Comment (gsheni) [inline docs/source/release_notes.rst:19]: If this is a breaking change, let's document it here.

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py:54]: I've mapped each case above to it's corresponding test
- `init(**params)`: `test_init_detect_partial_schema`
- `init(schema=partial_schema, **overriding_params)`: `test_init_with_partial_schema_override_schema`
- `init(schema=full_schema)`: `test_init_with_full_schema`
- `init(schema=full_schema, **ignored_params)`: `test_accessor_with_schema_parameter_warning`
- `init(schema=schema_with_columns_not_in_df)`:  `test_check_partial_schema`

If you'd like I can switch all the `init` tests to be either `'init_with_partial_schema` or `init_with_full_schema` tests although keeping them the way makes sure there are no breaking changes

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py]: I've made an alias for ColumnName

File: docs/source/release_notes.rst
Comment (gsheni) [inline docs/source/release_notes.rst:22]: @tuethan1999 Can you expand on this? We need to explain the behavior before the change, and the new behavior now. See [here](https://woodwork.alteryx.com/en/latest/release_notes.html#breaking-changes) for an example

File: docs/source/release_notes.rst
Comment (gsheni) [inline docs/source/release_notes.rst:26]:
```suggestion
    would error. Now, passing a partial schema to the `init` method calls
```

**Comment by tuethan1999** [comment]
@tamargrey now that we've already crossed the breaking change boundary, does it still make sense that we ignore arguments along with a full schema? It actually seems like the EvalML use case wants to pass in a full schema and specify which columns to reinfer. Tbh,`init_with_full_schema` seems more of an internal call to me.

File: woodwork/tests/accessor/test_table_accessor.py
Comment (tamargrey) [reply woodwork/tests/accessor/test_table_accessor.py:2594]: @tuethan1999 This test still needs to be updated to not just add a new element to the metadata dictionary, but to update the value of an element already in the dictionary. 

```python
sample_df.ww.init(table_metadata={'test': [1,2]})

...

test_df.ww._schema.metadata['test'].append(3)
```

That way it's checking that a deep copy has occured

**Review by tamargrey — CHANGES_REQUESTED** [review]
At this point, I think I'm with you on making `init` use partial schema init and allow overriding even with a full schema. As long as @gsheni is okay with that added breaking change, I say go for it.

> Tbh,`init_with_full_schema` seems more of an internal call to me.

I could see instances where you are using woodwork and are performing some operation that shouldn't invalidate the schema but will create a new DataFrame. Those cases might want the rigidity of `init_with_full_schema` to confirm no unexpected changes happened. I don't see much harm in leaving it public, especially since it could be the only remnant of the former init with schema behavior.

Though if we do make it private, we can also make `init_with_partial_schema` private and just expose `init`

**Comment by gsheni** [comment]
@tuethan1999 @tamargrey I am okay with the breaking change.

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py]: fixed!

**Review by freddyaboulton — COMMENTED** [review]
@tuethan1999 I think this breaks `ww.to_frame()` for pandas series?

I installed this branch locally and calling `to_frame()` on any ww series throws an error:

image

I really like what's going on in this branch though. I was able to update our Imputer, TargetEncoder, PCA, LDA, and DFSTransformer with this new init implementation for partial schemas rather than the evalml util that tried to do the same thing and I can get our unit tests to pass! (Thank you @tamargrey for the help in answering my questions). The thing about `to_frame` is the only "bug" I've seen with this.

I'll continue to chug along and report back!

**Comment by tamargrey** [comment]
> I installed this branch locally and calling to_frame() on any ww series throws an error:

@freddyaboulton @tuethan1999 Interestingly, there is one way where `to_frame` doesn't error for me locally: when the Series is from a DataFrame.

```
y = sample_df.ww['id']
y.ww.to_frame()
```

However `test_series_methods_returning_frame` also doesn't error.

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:1048]: Nice!

**Comment by tuethan1999** [comment]
@tamargrey @freddyaboulton thanks for bringing this bug up. Turns out it breaks when the series doesn't have a name! working on a fix

File: woodwork/table_accessor.py
Comment (davesque) [inline woodwork/table_accessor.py:119]: By doing `from __future__ import annotations` at the top as was mentioned in a previous review comment, you can use `str | LogicalType | None` instead of `Union[str, LogicalType, None]`.  We should probably adopt that convention for all union type annotations.

File: woodwork/table_accessor.py
Comment (davesque) [inline woodwork/table_accessor.py:122]: Similar to the earlier comment about `from __future__ import annotations` and being able to use `dict[...]`, you can do the same here with `list` and set`.  So `List[str]` and `Set[str]` would become `list[str]` and `set[str]` respectively.

File: woodwork/table_accessor.py
Comment (davesque) [inline woodwork/table_accessor.py:128]: Relating to some of the previous comments, for annotations like this where the explicit `Optional[...]` is necessary, it is equivalent and probably more concise to do the `from __future__ import annotations` as in previous comments and then do `bool | None` instead of `Optional[bool]`.

**Review by davesque — COMMENTED** [review]
Awesome to see some type annotations here!  Also great that you identified the proper generics to use such as `Iterable[...]`, etc.  Had some comments about some further improvements that are possible with the annotations.

File: woodwork/table_accessor.py
Comment (tamargrey) [reply woodwork/table_accessor.py:42]: yes, they're anything that can be a pandas column name (or for Dask or Koalas, what can be a dask or koalas column name)

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:62]: image

Formatting here and above isn't coming through in the docs build

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:135]: After reading through @davesque's comments I think these changes should be possible to do to let us remove the `Dict` and `Union` imports (unless I'm misunderstanding where we can use `dict` vs `Dict`)

```suggestion
                                 schema: TableSchema = None,
                                 index: str = None,
                                 time_index: str = None,
                                 logical_types: dict[ColumnName, str | LogicalType | None] = None,
                                 already_sorted: bool = False,
                                 name: str = None,
                                 semantic_tags: dict[ColumnName, str | list[str] | set[str]] = None,
                                 table_metadata: dict = None,
                                 column_metadata: dict[ColumnName, dict] = None,
                                 use_standard_tags: bool | UseStandardTagsDict = None,
                                 column_descriptions: dict[ColumnName, str] = None,
                                 column_origins: str | dict[ColumnName, str] = None,
                                 validate: bool = True,
                                 **kwargs) -> None:
```

(can also make similar changes in `_infer_missing_logical_types` and `_merge_use_standard_tags` )

**Review by tamargrey — COMMENTED** [review]
Looking good! Just want to make sure we're starting out our usage of python's typing really consistently!

File: woodwork/table_accessor.py
Comment (davesque) [reply woodwork/table_accessor.py:135]: @tamargrey Actually, @tuethan1999 and I realized that it might be a bit too early to adopt those next gen type annotation features.  I think `Union` and `Dict` still make sense at this point.  Otherwise, we might end up having to use python 3.9 to do the type checking with mypy.  That might be okay, but it could just as well introduce some complexities to our CI pipeline that didn't exist before.  It might also introduce subtle issues that come up as a result of the version mismatch.  Not sure how that would look but it seems possible.

Comment (tamargrey) [reply woodwork/table_accessor.py:135]: Ah, thanks for the explanation! I agree that being safe here makes sense!

File: woodwork/table_accessor.py
Comment (tuethan1999) [reply woodwork/table_accessor.py]: The formatting should be fixed! Is it ok if I keep the kwargs as` **kwargs`?

Comment (tamargrey) [reply woodwork/table_accessor.py]: looks good! Also I'm fine keeping the kwargs as they are in `init`!

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:42]: this needs to be moved to be right below `from woodwork.type_sys.utils import _is_numeric_series, col_is_datetime`

**Review by tamargrey — APPROVED** [review]

**Comment by tamargrey** [comment]
lgtm! Idk if anyone else wants to take a last look before merging, though

**Pull request merged**
